torch/csrc/distributed/c10d/c10d_frontend.cpp

```diff
@@ -0,0 +1,635 @@
+  } else {
+    int64_t group_dst_rank = getGroupRank(group, dst);
+    opts.rootRank = group_dst_rank;
+    work = group->reduce(tensors, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGatherMultiGPU(
+    std::vector<std::vector<at::Tensor>>& output_tensor_lists,
+    std::vector<at::Tensor>& input_tensor_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto work = group->allgather(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGather(
+    std::vector<at::Tensor>& tensor_list,
+    at::Tensor tensor,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  std::vector<std::vector<at::Tensor>> output_tensor_lists = {std::move(tensor_list)};
+  std::vector<at::Tensor> input_tensor_list = {std::move(tensor)};
+  auto work = group->allgather(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allGatherCoalesced(
+    std::vector<std::vector<at::Tensor>>& output_tensor_lists,
+    std::vector<at::Tensor>& input_tensor_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto work =
+      group->allgather_coalesced(output_tensor_lists, input_tensor_list);
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::gather(
+    at::Tensor tensor,
+    const std::vector<at::Tensor>& gather_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    int64_t dst,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  auto my_rank = getRank(default_pg_);
+
+  // TODO TODO how to implement this?
+  // _validate_output_list_for_rank(my_rank, dst, gather_list)
+  std::vector<std::vector<at::Tensor>> output_tensors;
+  if (dst == my_rank) {
+    output_tensors.push_back(gather_list);
+  }
+
+  std::vector<at::Tensor> input_tensors = {std::move(tensor)};
+
+  GatherOptions opts;
+  opts.rootRank = dst;
+
+  std::shared_ptr<ProcessGroup::Work> work;
+  if (group == default_pg_) {
+    work = group->gather(output_tensors, input_tensors, opts);
+  } else {
+    int64_t group_dst_rank = getGroupRank(group, dst);
+    opts.rootRank = group_dst_rank;
+    work = group->gather(output_tensors, input_tensors, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::scatter(
+    at::Tensor tensor,
+    std::vector<at::Tensor>& scatter_list,
```

Comment:
shouldn't all the vector<Tensor> argument in the APIs be const reference?

Reply:
Actually most of them cannot be const references, many times they are passed directly to methods in ProcessGroup, which doesn't take const reference:

https://github.com/pytorch/pytorch/blob/c1141b6f68dc7a842f3d947ffb3c9f261940d31d/torch/lib/c10d/ProcessGroup.hpp#L237